cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskTracker.java

```diff
@@ -4,51 +4,149 @@
+    dropwizardMetricRegistry.register(MetricRegistry.name(metricName, GAUGE_ONGOING_EXECUTION_IN_NON_KAFKA_ASSIGNER_MODE),
+                                      (Gauge<Integer>) () -> !_isKafkaAssignerMode
+                                                             && !inExecutionTasks(TaskType.cachedValues()).isEmpty() ? 1 : 0);
   }
 
   /**
-   * Get the execution task with replica action set with the given task state.
-   * @param taskState the task state to get the execution task for.
-   * @return the replica action execution task set with the given task state.
+   * Update the execution state of the task.
+   *
+   * @param task The task to update.
+   * @param newState New execution state of the task.
    */
-  public Set<ExecutionTask> taskForReplicaAction(ExecutionTask.State taskState) {
-    return _replicaActionTasks.get(taskState);
+  public void markTaskState(ExecutionTask task, State newState) {
+    _tasksByType.get(task.type()).get(task.state()).remove(task);
+    switch (newState) {
+      case PENDING:
+        break; // Let it go.
+      case IN_PROGRESS:
+        task.inProgress(_time.milliseconds());
+        updateDataMovement(task);
+        break;
+      case ABORTING:
+        task.abort();
+        break;
+      case ABORTED:
+        task.aborted(_time.milliseconds());
+        updateDataMovement(task);
+        break;
+      case COMPLETED:
+        task.completed(_time.milliseconds());
+        updateDataMovement(task);
+        break;
+      case DEAD:
+        task.kill(_time.milliseconds());
+        updateDataMovement(task);
+        break;
+      default:
+        break;
+    }
+    _tasksByType.get(task.type()).get(newState).add(task);
+  }
+
+  private void updateDataMovement(ExecutionTask task) {
+    if (task.type() == TaskType.LEADER_ACTION) {
+      return;
+    }
+    long dataToMove = task.proposal().dataToMoveInMB();
+    if (task.state() == State.IN_PROGRESS) {
+      _remainingDataToMoveInMB -= dataToMove;
+      _inExecutionDataMovementInMB += dataToMove;
+    } else if (task.state() == State.ABORTED || task.state() == State.DEAD || task.state() == State.COMPLETED) {
+      _inExecutionDataMovementInMB -= dataToMove;
+      _finishedDataMovementInMB += dataToMove;
+    }
   }
 
   /**
-   * Get the execution task with leader action set with the given task state.
-   * @param taskState the task state to get the execution task for.
-   * @return the leader action execution task set with the given task state.
+   * Add new tasks to ExecutionTaskTracker to track their execution state.
+   *
+   * @param tasks    New tasks to add.
+   * @param taskType Task type of new tasks.
    */
-  public Set<ExecutionTask> taskForLeaderAction(ExecutionTask.State taskState) {
-    return _replicaActionTasks.get(taskState);
+  public void initializeTask(Collection<ExecutionTask> tasks, TaskType taskType) {
```

Comment:
Since [this method](https://github.com/linkedin/cruise-control/blob/90e0e3f4a2e4044718badf9661e90bb18c0ce9a0/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java#L163) returns collections...